apps/aechannel/src/aesc_fsm.erl

```diff
@@ -2292,23 +2290,113 @@ new_onchain_tx(channel_withdraw_tx, #{acct := ToId,
+                               responder := Responder}
+                    , state       = State } = D, CurrHeight) ->
+    Account = my_account(D),
+    TTL = adjust_ttl(maps:get(ttl, Opts, 0)),
+    {_Round, SignedTx} = aesc_offchain_state:get_latest_signed_tx(State),
+    Payload = case aesc_utils:is_offchain_tx(SignedTx) of
+                  true ->
+                      aetx_sign:serialize_to_binary(SignedTx);
+                  false ->
+                      %% on-chain tx, we assume included in the chain
+                      <<>>
+              end,
+    {ok, Poi} = aesc_offchain_state:poi([{account, Initiator},
+                                         {account, Responder}], State),
+    Opts1 =
+          Opts#{ channel_id => aeser_id:create(channel, ChanId)
+               , from_id    => aeser_id:create(account, Account)
+               , payload    => Payload
+               , poi        => Poi
+               , ttl        => TTL},
+    {ok, Tx} = new_onchain_tx_(aesc_close_solo_tx, Opts1, CurrHeight),
+    {ok, Tx, []};
+new_onchain_tx(channel_slash_tx, Opts,
+               #data{ on_chain_id = ChanId
+                    , opts = #{initiator := Initiator,
+                               responder := Responder}
+                    , state       = State} = D, CurrHeight) ->
+    Account = my_account(D),
+    Def = tx_defaults(channel_slash_tx, #{acct => Account}, D),
+    Opts1 = maps:merge(Def, Opts),
+    TTL = adjust_ttl(maps:get(ttl, Opts1, 0)),
+    {ok, Poi} = aesc_offchain_state:poi([{account, Initiator},
+                                         {account, Responder}], State),
+    Opts2 =
+        Opts1#{ channel_id => aeser_id:create(channel, ChanId)
+              , from_id    => aeser_id:create(account, Account)
+              , poi        => Poi
+              , ttl        => TTL},
+    {ok, Tx} = new_onchain_tx_(aesc_slash_tx, Opts2, CurrHeight),
+    {ok, Tx, []}.
+
+
+
+
+new_onchain_tx_(Mod, Opts, CurrHeight) ->
+    case maps:is_key(fee, Opts) of
+        true -> % use preset fee
+            apply(Mod, new, [Opts]);
+        false ->
+            create_with_minuimum_fee(Mod, Opts#{fee => 0}, CurrHeight)
+    end.
+
+create_with_minuimum_fee(Mod, Opts, CurrHeight) ->
+    create_with_minuimum_fee(Mod, Opts, CurrHeight, 5).
+
+create_with_minuimum_fee(_, _, _, Attempts) when Attempts < 1 ->
+    erlang:error(could_not_compute_fee);
+create_with_minuimum_fee(Mod, Opts, CurrHeight, Attempts) ->
+    {ok, Tx} = apply(Mod, new, [Opts]),
+    MinTxFee = aetx:min_fee(Tx, CurrHeight),
+    MinGas = aetx:min_gas(Tx, CurrHeight),
+    MinMinerGasPrice = aec_tx_pool:minimum_miner_gas_price(),
+    MinGasRequirements = MinGas * MinMinerGasPrice,
+    MinFee = max(MinTxFee, MinGasRequirements),
+    TxFee = aetx:fee(Tx),
+    case MinFee =< TxFee of
+        true ->
+            {ok, Tx};
+        false ->
+            create_with_minuimum_fee(Mod, Opts#{fee => MinFee}, CurrHeight,
```

Comment:
I see the only argument that changes in the recursion is the fee. I understand this is because a larger fee implies a larger serialization.

So 5 attempts seems enough.

Reply:
Yes, I don't see how it could possibly go beyond 3 attempts, but put 5 anyway